Core implementation

mykhailonakko opened the pull request:
Created a framework structure:
- added necessary dependencies.
- added Report Portal listener.

mykhailonakko committed:
Framework structure added

added:
-properties
-testng and junit runners
-reportportal listener
-dummy test

(cherry picked from commit f8679155e93182cec3ba0fd0ad33455228c737c1)

mykhailonakko committed:
Update .gitignore

mykhailonakko committed:
Update .gitignore

mykhailonakko committed:
Merge branch 'master' into core_implementation

# Conflicts:
#	.gitignore

NovoDan commented on .gitignore:
```diff
@@ -0,0 +1,39 @@
+target/
+!.mvn/wrapper/maven-wrapper.jar
+!**/src/main/**/target/
+!**/src/test/**/target/
```
just one 'target/' would be enough

mykhailonakko replied:
Thanks, will fix that.

mykhailonakko replied:
fixed

NovoDan commented on .gitignore:
```diff
@@ -0,0 +1,39 @@
+!**/src/test/**/target/
+
+### IntelliJ IDEA ###
+.idea
```
same here

NovoDan commented on src/main/java/org/rptest/core/logger/JavaUtilLogger.java:
```diff
@@ -0,0 +1,31 @@
+        StackTraceElement[] stElements = Thread.currentThread().getStackTrace();
+        String className = stElements[3].getClassName();
+        try {
+            return Class.forName(className).getName();
```
Doesn't it will return same value as stored in 'className'?

mykhailonakko replied:
Good question. 

I tried with a few classes and it returned the same thing, but here's what chatGPT has to say about it:

_You're correct that in most cases, `stElements[3].getClassName();` and `Class.forName(className).getName();` will return the same value. The main reason for using `Class.forName(className)` in the try-catch block is to ensure that the class name obtained from the stack trace actually exists and can be loaded by the Java ClassLoader.

In some cases, the class might not be available at runtime due to various reasons such as class loading issues, classpath configuration problems, or the class being removed or relocated in the project. By using `Class.forName(className)`, the code checks if the class can be loaded successfully. If a `ClassNotFoundException` is thrown, it indicates that the class could not be found or loaded, and an error message is logged accordingly.

However, if you are confident that the class will always be available at runtime, you can simplify the method by removing the try-catch block and returning the class name directly from the stack trace element_

```java
private String getCallerClassName() {
    StackTraceElement[] stElements = Thread.currentThread().getStackTrace();
    return stElements[3].getClassName();
}
```

_This simplified version of the method assumes that the class name obtained from the stack trace will always be valid and loadable. Keep in mind that if there's any issue with loading the class, this simplified version will not log an error message and will not handle the potential `ClassNotFoundException`._

ihor-dynka commented:
in pom.xml all dependencies and plugins version move to properties section. It increases readability and adds option to be able to manage version via command line

ihor-dynka commented on src/main/java/org/rptest/core/logger/JavaUtilLogger.java:
```diff
@@ -0,0 +1,31 @@
+
+    private String getCallerClassName() {
+        StackTraceElement[] stElements = Thread.currentThread().getStackTrace();
+        String className = stElements[3].getClassName();
```
please avoid magic numbers

mykhailonakko committed:
Test Runner module

-added a login test
-added browser implementations and handlers
-added Ui element implementations and handlers
-added Login page object
-added a CSV file handler
-added a csv file with test data

mykhailonakko committed:
POM and XML refactoring

- removed redundant lines in POM file
- provided variables for dependency versions in POM file

ihor-dynka commented on src/main/java/org/rptest/core/browser/PwBrowser.java:
```diff
@@ -0,0 +1,45 @@
+import static com.microsoft.playwright.BrowserType.LaunchOptions;
+import static org.rptest.core.config.Property.COMMON_PROPERTY;
+
+public class PwBrowser {
```
avoid abbreviation, use full names

ihor-dynka commented on src/main/java/org/rptest/core/elements/UiElementFactory.java:
```diff
@@ -0,0 +1,29 @@
+    private static final UiLibrary library = Enum.valueOf(UiLibrary.class, Property.COMMON_PROPERTY.uiLibrary().toUpperCase());
+
+    public static UiElement findElementById(String id) {
+        switch (library) {
```
this login should be executed once

ihor-dynka commented on src/main/java/org/rptest/core/utilities/FileUtils.java:
```diff
@@ -0,0 +1,28 @@
+@UtilityClass
+public class FileUtils {
+    public static List<CSVRecord> readCSV(String filePath) throws IOException {
+        Reader reader = new FileReader(filePath);
```
another option can be using try with resources and rethrow another exception

ihor-dynka commented:
- please add option to run the tests in parallel mode (investigate the difference between class and methods parallelization, forks vs threads)

ihor-dynka commented:
- add branch policies to allow merging only after at least 1 approve and all comments resolved

mykhailonakko committed:
Parallel test run implementation

Added JUnit 5 tests.
Refactored Login page.

mykhailonakko committed:
Deleted UiElementFactory as logic has changed and browsers are responsible for finding elements

mykhailonakko committed:
Minor refactoring

mykhailonakko committed:
Switched the properties back to localhost

removed redundant tests.

ihor-dynka committed:
Merge pull request #2 from mykhailonakko/parallel_execution_implementation

Parallel execution implementation

mykhailonakko committed:
Removed dead code and commented lines
